### ISSUE TITLE
Fatal error with class invariants

### ISSUE THREAD

**Issue by jad-hamza** [epfl-lara/leon#197]
```scala
import leon.lang._

object Test  {
  
  case class DefinedAtZero(m: Map[BigInt,BigInt])  {
    require(m.contains(0))
        
  }

  def receive(m: DefinedAtZero) = {
    m 
  } ensuring(res => res.m(0) == 5)
}
```

I obtain the empty map as a counter-example to function "receive" (not a valid counter-example, because it's not defined at zero)
Using option --checkmodels produces a fatal error

> [Warning ] - Model leads to evaluation error: Stack overflow
> [ Error  ] Something went wrong. The model should have been valid, yet we got this: Model(
> [ Error  ]   m                    -> DefinedAtZero(Map[BigInt,BigInt](Map()))
> [ Error  ] ) for formula m.m(BigInt(0)) != BigInt(5)
> [Warning ]  => UNKNOWN
> [ Fatal  ] There were errors.

Is this a known issue?
Thanks

**Comment by samarion** [comment]
Yes, this is an issue with the extraction of maps from SMT models. There is unfortunately a slight discrepancy between the representation of maps in SMT and in Scala, and this is visible during model extraction. We're thinking of introducing some form of inductive datatypes that could solver this particular issue.

**Comment by samarion** [comment]
Fixed stack overflow issue in 5463d3f1e0fc007a6dc349a51706c3d758c9fbe4.

**Comment by jad-hamza** [comment]
Thanks!
